### PR TITLE
Add support for calculating taxes by sync webhooks

### PR DESCRIPTION
**Pull request by stnatic**
This PR adds functionality of calculating taxes by sync webhooks.

Prices are cached (indefinitely for orders, or for `settings.CHECKOUT_PRICES_TTL` for checkouts), until they are
invalidated by mutations.

Instead of using prices directly from the database, now all prices need to be fetched using functions from 
`order.calculations`/`checkout.calculations` modules

Order and checkout payloads are extended with product and product type metadata for each line. 
Inside those, tax code will be saved for the key `f'{app.identifier}.code'` (if configured in the dashboard.)

New permissions:
* `Checkout.HANDLE_TAXES`

Response payload schema for tax apps:

```python3
@dataclass
class TaxLineData:
    id: int
    currency: str
    tax_rate: Decimal
    unit_net_amount: Decimal
    unit_gross_amount: Decimal
    total_gross_amount: Decimal
    total_net_amount: Decimal


@dataclass
class TaxData:
    currency: str
    total_net_amount: Decimal
    total_gross_amount: Decimal
    subtotal_net_amount: Decimal
    subtotal_gross_amount: Decimal
    shipping_price_gross_amount: Decimal
    shipping_price_net_amount: Decimal
    shipping_tax_rate: Decimal
    lines: List[TaxLineData]
```

```mermaid
sequenceDiagram
    User ->>+ Saleor: Query requesting model's prices
    opt if prices are invalid
        Saleor->>+Manager: Calculate taxes for model <br>with old approach
        loop iterate over tax plugins
          Manager ->>+ Tax Plugin: Fetch prices by manager.calculate_*
          Tax Plugin -->>- Manager: Return prices (with optional taxes)
        end
        Manager -->>- Saleor: Apply prices to model
        Saleor ->>+ Manager: Calculate taxes <br>with new approach
          Manager ->>+ Webhook Plugin: Fetch taxes by <br>manager.get_taxes_for_*
          loop iterate over tax apps
            Webhook Plugin ->>+ Tax App: Send sync webhook
            Tax App -->>- Webhook Plugin: Send tax data
            Webhook Plugin ->> Webhook Plugin: Exit when received valid tax data<br>or no more apps
          end
          Webhook Plugin -->>- Manager: Return Optional[TaxData]
        opt if received tax data
          Manager-->>-Saleor: Apply taxes to model
        end
        Saleor ->> Saleor: Save all model changes to DB
    end
    Saleor -->>- User: Return model's prices
```

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/webhook/payloads.py [hunk 1]
```diff
@@ -1069,3 +1160,176 @@ def generate_excluded_shipping_methods_for_checkout_payload(
...
+    }
+
+
+def generate_order_payload(
+    order: "Order",
+    requestor: Optional["RequestorOrLazyObject"] = None,
+    with_meta: bool = True,
+):
+    manager = get_plugins_manager()
+    lines = OrderLine.objects.select_related("variant__product__product_type")
```
Comment (stnatic) [inline saleor/webhook/payloads.py:1214]:
```suggestion
    lines = order.lines.select_related("variant__product__product_type")
```

File: saleor/checkout/migrations/0041_auto_20220329_1612.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+
+def set_default_checkout_line_currency(apps, schema_editor):
+    CheckoutLine = apps.get_model("checkout", "CheckoutLine")
+    lines = CheckoutLine.objects.select_related("checkout")
+
+    for line in lines:
+        line.currency = line.checkout.currency
+
+    CheckoutLine.objects.bulk_update(lines, ["currency"])
```
Comment (stnatic) [inline saleor/checkout/migrations/0041_auto_20220329_1612.py:16]: @fowczarek @cmiacz We'd like to hear your opinion on performance of this piece.

File: CHANGELOG.md [hunk 3]
```diff
@@ -158,6 +185,13 @@ All notable, unreleased changes to this project will be documented in this file.
   `CheckoutAddPromoCode`, `CheckoutPaymentCreate` will raise a ValidationError when product in the checkout is
   unavailable - #8978 by @IKarbowiak
 - Add `withChoices` flag for Attribute type - #7733 by @dexon44
+`CheckoutAddPromoCode`, `CheckoutPaymentCreate` will raise a ValidationError when product in the checkout is
```
Comment (korycins) [inline CHANGELOG.md:188]: It looks like something that is not the scope of this PR.

File: CHANGELOG.md [hunk 4]
```diff
@@ -115,6 +115,33 @@ All notable, unreleased changes to this project will be documented in this file.
 
 - Don't run plugins when calculating checkout's total price for available shipping methods resolution - #9121 by @rafalp
   - Use either net or gross price depending on store configuration.
+- Update webhook payload for `list_shipping_methods_for_checkout`
```
Comment (korycins) [inline CHANGELOG.md:118]: Please add a reference number to PR. As it is done for the rest of the records

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 repos:
   - repo: https://github.com/ambv/black
-    rev: 22.1.0
+    rev: 22.3.0
```
Comment (korycins) [inline .pre-commit-config.yaml:3]: Is it expected to change? Not sure if this should be palced in this PR.

File: saleor/checkout/migrations/0041_auto_20220329_1612.py (hunk 2 above)
Comment (korycins) [reply saleor/checkout/migrations/0041_auto_20220329_1612.py:16]: IMO there is a risk of OOM  here.  Also not sure if we shouldn't have a separate migration for populating this field

File: saleor/checkout/models.py [hunk 6]
```diff
@@ -186,6 +239,100 @@ class CheckoutLine(models.Model):
...
+    undiscounted_unit_price = TaxedMoneyField(
+        net_amount_field="undiscounted_unit_price_net_amount",
+        gross_amount_field="undiscounted_unit_price_gross_amount",
+    )
+    unit_price_net_amount = models.DecimalField(
+        max_digits=settings.DEFAULT_MAX_DIGITS,
+        decimal_places=settings.DEFAULT_DECIMAL_PLACES,
+        default=0,
+    )
+    unit_price_gross_amount = models.DecimalField(
```
Comment (korycins) [inline saleor/checkout/models.py:265]: Would be better to have default=Decimal(0) here. Function like quantitze_price could fail as they will get int as an input. 
Also, question. Are you sure that we should set default values for these fields? I am not sure of this.

File: saleor/checkout/migrations/0041_auto_20220329_1612.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+    operations = [
+        migrations.AlterModelOptions(
+            name="checkout",
+            options={
+                "ordering": ("-last_change", "pk"),
+                "permissions": (
+                    ("manage_checkouts", "Manage checkouts"),
+                    ("handle_checkouts", "Handle checkouts"),
+                    ("handle_taxes", "Handle taxes"),
```
Comment (korycins) [inline saleor/checkout/migrations/0041_auto_20220329_1612.py:33]: Adding new permission requires a separate migration that will assign given perm to all groups that already have an access to manager_checkouts. See migration of handle_checkouts as an example. https://github.com/saleor/saleor/pull/9402

File: saleor/checkout/migrations/0041_auto_20220329_1612.py (hunk 2 above)
Comment (fowczarek) [inline saleor/checkout/migrations/0041_auto_20220329_1612.py:16]: Maybe?
```suggestion
    CheckoutLine.objects.update(currency=F("checkout__currency"))
```

**Review by fowczarek — APPROVED** [review]

File: saleor/graphql/order/mutations/orders.py [hunk 8]
```diff
@@ -225,6 +229,17 @@ def get_instance(cls, info, **data):
...
         return instance
 
+    @classmethod
+    def should_invalidate_prices(cls, instance, cleaned_input, new_instance) -> bool:
+        return any(
+            cleaned_input.get(field) is not None
+            for field in [
+                "shipping_address",
+                "billing_address",
+                "shipping_method",
```
Comment (korycins) [inline saleor/graphql/order/mutations/orders.py:239]: Is it correct? I don't see shipping_method in input for this method.

File: saleor/checkout/migrations/0041_auto_20220329_1612.py (hunk 2 above)
Comment (mateuszgrzyb) [reply saleor/checkout/migrations/0041_auto_20220329_1612.py:16]: Also, another alternative would be:
```python3
CheckoutLine.objects.update(
    currency=Subquery(
        Checkout.objects.filter(pk=OuterRef("checkout")).values("currency")[:1]
    )
)
```
or raw SQL operation.
They should be equivalent to update with `F` expression.

File: saleor/checkout/migrations/0041_auto_20220329_1612.py (hunk 7 above)
Comment (mateuszgrzyb) [reply saleor/checkout/migrations/0041_auto_20220329_1612.py:62]: Resolved in recent commit.

File: CHANGELOG.md (hunk 3 above)
Comment (mateuszgrzyb) [reply CHANGELOG.md:196]: Resolved in recent commit.

File: CHANGELOG.md (hunk 4 above)
Comment (mateuszgrzyb) [reply CHANGELOG.md]: Resolved in recent commit.

File: saleor/graphql/order/mutations/orders.py (hunk 8 above)
Comment (mateuszgrzyb) [reply saleor/graphql/order/mutations/orders.py]: Resolved in recent commit.

File: saleor/checkout/models.py (hunk 6 above)
Comment (mateuszgrzyb) [reply saleor/checkout/models.py:271]: Why not? All checkout prices will be calculated by manager methods, I don't think we should duplicate those calculations when eg. creating a new line.

File: saleor/checkout/migrations/0041_auto_20220329_1612.py (hunk 2 above)
Comment (fowczarek) [reply saleor/checkout/migrations/0041_auto_20220329_1612.py:17]: Solved in 
https://github.com/saleor/saleor/pull/9526/commits/43bcac3eee309d29660cdbea3a0fb7ece2a5fb47

**Pull request closed without merging**